### PR TITLE
Add the field 'nodeName' in the HTTP header of the rest Router

### LINKED ISSUE #4754: 使用Router特性将消息从边缘节点发送到云端时，不知道消息来自哪个节点
**What would you like to be added/modified**:
希望能够在使用Router特性，将消息从边缘端发送云端时，在rest消息头中添加消息来源的节点名称
修改位置：
`github.com/kubeedge/kubeedge/cloud/pkg/router/provider/rest/rest.go`

image

**nodeName, _ := data["nodeName"].(string)**
**req, err := httpUtils.BuildRequest(http.MethodPost, r.Endpoint, bytes.NewReader(content), "", nodeName)**

**Why is this needed**:
有些场景，云端rest服务在接收到消息之后，希望知道消息来源于哪个节点，以便做后续处理。

### PR DESCRIPTION
**Pull request by kelpliu**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
有些场景，云端rest服务在接收到消息之后，希望知道消息来源于哪个节点，以便做后续处理。

**Which issue(s) this PR fixes**:

Fixes #4754 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: cloud/pkg/router/provider/rest/rest.go [hunk 1]
```diff
@@ -201,7 +201,8 @@ func (r *Rest) GoToTarget(data map[string]interface{}, stop chan struct{}) (inte
 	if !ok || len(content) == 0 {
 		return nil, errors.New("invalid convert to []byte")
 	}
-	req, err := httpUtils.BuildRequest(http.MethodPost, r.Endpoint, bytes.NewReader(content), "", "")
+	nodeName, _ := data["nodeName"].(string)
```
Comment (vincentgoat) [inline cloud/pkg/router/provider/rest/rest.go:204]: Type confusion is shown here, this should be avoided by checking for an error when type casting.
```suggestion
	nodeName, ok := data["nodeName"].(string)
	if !ok {...
```

Comment (kelpliu) [reply cloud/pkg/router/provider/rest/rest.go]: ok

**Comment by kelpliu** [comment]
The CI  e2e(websocket) failed,    maybe the   eventBus to rest or  serviceBus to rest   tests  allow having  invalid nodeName。

**Comment by vincentgoat** [comment]
> The CI e2e(websocket) failed, maybe the eventBus to rest or serviceBus to rest tests allow having invalid nodeName。 Can I add default value like nodeName = ""

As you modified before, ci also needs to synchronize the update.

**Comment by kelpliu** [comment]
> > The CI e2e(websocket) failed, maybe the eventBus to rest or serviceBus to rest tests allow having invalid nodeName。 Can I add default value like nodeName = ""
> 
> As you modified before, ci also needs to synchronize the update.

Sorry, I  do not know how  to modify the tests correctly:  E2E_CREATE_RULE_2, E2E_CREATE_RULE_4.      could you please give some guidance or guidance link.

File: cloud/pkg/router/provider/rest/rest.go [hunk 2]
```diff
@@ -201,7 +201,13 @@ func (r *Rest) GoToTarget(data map[string]interface{}, stop chan struct{}) (inte
 	if !ok || len(content) == 0 {
 		return nil, errors.New("invalid convert to []byte")
 	}
-	req, err := httpUtils.BuildRequest(http.MethodPost, r.Endpoint, bytes.NewReader(content), "", "")
+	nodeName, ok := data["nodeName"].(string)
+	if !ok {
+		err := fmt.Errorf("input data does exist valid value \"nodeName\"")
```
Comment (vincentgoat) [inline cloud/pkg/router/provider/rest/rest.go:206]: For compatibility reasons, it is fine to modify this.
But the log should be `input data does not exist valid value \"nodeName\"`?

Comment (kelpliu) [reply cloud/pkg/router/provider/rest/rest.go]: ok

File: cloud/pkg/router/provider/rest/rest.go [hunk 3]
```diff
@@ -201,7 +201,13 @@ func (r *Rest) GoToTarget(data map[string]interface{}, stop chan struct{}) (inte
 	if !ok || len(content) == 0 {
 		return nil, errors.New("invalid convert to []byte")
 	}
-	req, err := httpUtils.BuildRequest(http.MethodPost, r.Endpoint, bytes.NewReader(content), "", "")
+	nodeName, ok := data["nodeName"].(string)
+	if !ok {
+		err := fmt.Errorf("input data does not exist valid value \"nodeName\"")
+		klog.Warningf(err.Error())
+		nodeName = ""
```
Comment (fisherxu) [inline cloud/pkg/router/provider/rest/rest.go:208]: If not ok, the `nodeName` will not be ""? Is it needed to set the "" value here? :)

**Review by fisherxu — COMMENTED** [review]
Thanks for the contributions! @kelpliu

File: cloud/pkg/router/provider/rest/rest.go (hunk 3 above)
Comment (kelpliu) [reply cloud/pkg/router/provider/rest/rest.go:208]: if  return err,  the  two E2E tests  failed  in  CI.

Comment (fisherxu) [reply cloud/pkg/router/provider/rest/rest.go:208]: > if return err, the two E2E tests failed in CI.

Yes, no need to return err here. And I mean it no need to set empty value to nodename, which was originally empty.

Comment (kelpliu) [reply cloud/pkg/router/provider/rest/rest.go:208]: Ooooh, I see.   if    "nodeName, ok := data["nodeName"].(string) "  convert error  , "nodeName"  default value is empty.

I  will modify the code and push.  Thank you.

**Comment by kelpliu** [comment]
> @kelpliu Please squash the five commits, thanks :)

ok.
but it  looks like  the unit test  controllermanager_test failed.   Is it my modification that caused the  failure?

**Review by fisherxu — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**